src/tsobj-file-system-provider.ts

```diff
@@ -0,0 +1,264 @@
+    this.ctime = Date.now();
+    this.mtime = Date.now();
+    this.size = 0;
+    this.name = name;
+  }
+}
+
+export class Directory implements vscode.FileStat {
+  type: vscode.FileType;
+  ctime: number;
+  mtime: number;
+  size: number;
+
+  name: string;
+  entries: Map<string, File | Directory>;
+
+  constructor(name: string) {
+    this.type = vscode.FileType.Directory;
+    this.ctime = Date.now();
+    this.mtime = Date.now();
+    this.size = 0;
+    this.name = name;
+    this.entries = new Map();
+  }
+}
+
+export type Entry = File | Directory;
+
+export class TSObjFileSystemProvider implements vscode.FileSystemProvider {
+  // Implementation of the `onDidChangeFile` event
+  onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = new vscode.EventEmitter<
+    vscode.FileChangeEvent[]
+  >().event;
+
+  watch(): vscode.Disposable {
+    throw new Error('Watch not supported');
+  }
+
+  stat(uri: vscode.Uri): vscode.FileStat | Thenable<vscode.FileStat> {
+    Logger.info(`stat: ${uri.toString()}`, 'tsobj-fsp');
+    const { hostname, resourcePath } = this.extractHostAndPath(uri);
+
+    const command = `ssh ${hostname} "stat -L -c '{\\"type\\": \\"%F\\", \\"size\\": %s, \\"ctime\\": %Z, \\"mtime\\": %Y}' ${resourcePath}"`;
+
+    return new Promise((resolve, reject) => {
+      exec(command, (error, stdout) => {
+        if (error) {
+          reject(error);
+        } else {
+          const result = JSON.parse(stdout.trim());
+          const type =
+            result.type === 'directory' ? vscode.FileType.Directory : vscode.FileType.File;
+          const size = result.size || 0;
+          const ctime = result.ctime * 1000;
+          const mtime = result.mtime * 1000;
+          resolve({ type, size, ctime, mtime });
+        }
+      });
+    });
+  }
+
+  readDirectory(uri: vscode.Uri): Promise<[string, vscode.FileType][]> {
+    Logger.info(`readDirectory: ${uri.toString()}`, 'tsobj-fsp');
+
+    const { hostname, resourcePath } = this.extractHostAndPath(uri);
+    Logger.info(`hostname: ${hostname}`, 'tsobj-fsp');
+    Logger.info(`remotePath: ${resourcePath}`, 'tsobj-fsp');
+
+    const command = `ssh ${hostname} ls -lA "${resourcePath}"`;
+    return new Promise((resolve, reject) => {
+      exec(command, (error, stdout) => {
+        if (error) {
+          reject(error);
+        } else {
+          const lines = stdout.trim().split('\n');
+          const files: [string, vscode.FileType][] = [];
+          for (const line of lines) {
+            // Skip the first line if it starts with 'total'
+            if (line.startsWith('total')) {
+              continue;
+            }
+            const parts = line.trim().split(/\s+/);
+            const name = parts[parts.length - 1];
+            const isDirectory = parts[0].startsWith('d');
+            const type = isDirectory ? vscode.FileType.Directory : vscode.FileType.File;
+            files.push([name, type]);
+          }
+          resolve(files);
```

Comment:
can we sort the files so that directories show up first? This is how the regular file explorer sorts the display